File: internal/limiter/bucket.go
package limiter

import (
	"sync"
	"time"
)

// tokenBucket is a token bucket implementation
// hold 'capacity' tokens, refilling at 'refillRate' tokens per second
// each request consumes 'cost' tokens
type tokenBucket struct {
	mu sync.Mutex
	tokens float64
	capacity float64
	refillRate float64
	lastRefill time.Time
	lastAccess time.Time
}

func newTokenBucket(capacity, refillRate float64) *tokenBucket {
	now := time.Now()

	return &tokenBucket{
		tokens: capacity,
		capacity: capacity,
		refillRate: refillRate,
		lastRefill: now,
		lastAccess: now,
	}
}

// Refill bucket based on elapsed time, attemp to deduct cost tokens. If successful, returns how many tokens remain. Otherwise, returns how long until refill.
func (b *tokenBucket) allow(cost float64) (allowed bool, remaining float64, retryAfter time.Duration) {
	b.mu.Lock()
	defer b.mu.Unlock()

	now := time.Now()
	elapsed := now.Sub(b.lastRefill).Seconds()
	b.tokens = minF(b.capacity, b.tokens+b.refillRate*elapsed)
	b.lastRefill = now
	b.lastAccess = now

	if b.tokens >= cost {
		b.tokens -= cost
		return true, b.tokens, 0
	}

	deficit := cost - b.tokens
	retryAfter = time.Duration(deficit/b.refillRate * float64(time.Second))
	return false, b.tokens, retryAfter
}


// Report if bucket has been untouched since cutoff
func (b *tokenBucket) idleSince(cutoff time.Time) bool {
	b.mu.Lock()
	defer b.mu.Unlock()

	return b.lastAccess.Before(cutoff)
}

func minF(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

File: app/bucket.py
"""
Token bucket implementation.

FastAPI route handlers are async def, so many asyncio coroutines interleaving on a single threat, not OS threads running in parallel like in Go's goroutines.
We use asyncio.Lock as it is the idiomatic equivalent to Go's mutex.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass

@dataclass
class AllowResult:
    allowed: bool
    remaining: int
    retry_after_seconds: float

class TokenBucket:
    """One bucket per client key."""

    __slots__ = ("_lock", "tokens", "capacity", "refill_rate", "last_refill", "last_access")

    def __init__(self, capacity: int, refill_rate: float) -> None:
        now = time.monotonic()
        self._lock = asyncio.Lock()
        self.tokens = capacity
        self.capacity = capacity
        self.refill_rate = refill_rate
        self.last_refill = now
        self.last_access = now

    async def allow(self, cost: float) -> AllowResult:
        """Refill based on elapsed time, then attempt to consume 'cost' tokens."""

        async with self._lock:
            now = time.monotonic()
            elapsed = now - self.last_access
            self.tokens = min(self.capacity, self.tokens + elapsed * self.refill_rate)
            self.last_refill = now
            self.last_access = now

            if self.tokens >= cost:
                self.tokens -= cost
                return AllowResult(allowed=True, remaining=self.tokens, retry_after_seconds=0.0)
            
            deficit = cost - self.tokens
            retry_after = deficit / self.refill_rate
            return AllowResult(allowed=False, remaining=self.tokens, retry_after_seconds=retry_after)
        
    async def idle_since(self, cutoff: float) -> bool:
        """Report if the bucket has been idle since cutoff."""
        async with self._lock:
            return self.last_access < cutoff